### PR TITLE
Adding a test for the null pointer exception I ran across today.

### PR DESCRIPTION
**Pull request by berlin-ab**
Should the constructor take another argument, or should the constructor
be more defensive?

### CODE REVIEW

**Comment by Crisfole** [comment]
Hey Adam, was this NPE from having an null in the `custom` field? I have a fix for that in place.

My goal was to have all constructors work as is...I'll go test that locally.

**Comment by berlin-ab** [comment]
Yep, that's the one. We also found another NPE when calling #log.

File: rollbar/src/test/java/com/rollbar/RollbarTest.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+public class RollbarTest {
+
+    @Test
+    public void itDoesNotThrowAnExceptionWhenConfiguringWithJustAccessTokenAndEnvironment() {
+        Boolean exceptionDidHappen = false;
+
+        try {
+            new Rollbar("some-access-token", "some-environment");
+        } catch (NullPointerException e) {
+            exceptionDidHappen = true;
```
Comment (Crisfole) [inline rollbar/src/test/java/com/rollbar/RollbarTest.java]: You can just use `fail()` here, and then drop the `exceptionDidHappen` field.

Also you might be better off with:

```java
Rollbar r = new Rollbar("Some access token", "environment name");
pass();
```

The uncaught error will propagate up and will display better in the test panels/command line. (Because it will include the exception stack trace).

**Comment by Crisfole** [comment]
@berlin-ab Please do file the other test! This is a new library that I'm really excited about, but it needs some battle hardening.

**Comment by berlin-ab** [comment]
I wanted to make the intention of the test clear and I wasn't sure that raising the exception would do that. I'll try out `fail()` and see how I feel about it.

**Comment by berlin-ab** [comment]
Thanks for making this library.  I've used Rollbar on other projects in different languages, and I was disappointed to not find something more stable for Java.  I'm happy to help.

**Comment by Crisfole** [comment]
@berlin-ab I've also got some `hamcrest` test utilities already present. You can probably find something like 'doesNotThrow' (maybe).

**Comment by Crisfole** [comment]
Better yet, you could probably just create a `testContructorWorks` test:

```java
Rollbar r = new Rollbar("at", "env");
assertEqual("at", r.accessToken());
assertEqual("env", r.environment());
```

**Comment by berlin-ab** [comment]
Ok, updated with production code. 

Side note, I'm not exactly sure what `custom` means for this variable. Can you think of another name that is more descriptive?

File: rollbar/src/main/java/com/rollbar/Rollbar.java [hunk 2]
```diff
@@ -984,4 +986,12 @@ protected void setTransform(PayloadTransform transform) {
     public Rollbar transform(PayloadTransform transform) {
         return new Rollbar(accessToken, environment, sender, codeVersion, platform, language, framework, context, request, person, server, custom, notifier, responseHandler, filter, transform);
     }
+
+    private LinkedHashMap<String, Object> wrapCustom(Map<String, Object> custom) {
+        return new LinkedHashMap<String, Object>(custom);
+    }
+
+    private LinkedHashMap<String, Object> defaultCustom() {
+        return new LinkedHashMap<String, Object>();
```
Comment (Crisfole) [inline rollbar/src/main/java/com/rollbar/Rollbar.java]: In this case the 'defaultCustom' would actually need to be null, which would cause it to be skipped in the serializer.

File: rollbar/src/main/java/com/rollbar/Rollbar.java [hunk 3]
```diff
@@ -94,7 +96,7 @@ public Rollbar(String accessToken, String environment, Sender sender, String cod
         this.request = request;
         this.person = person;
         this.server = server;
-        this.custom = new LinkedHashMap<String, Object>(custom);
+        this.custom =  custom == null ? defaultCustom() : wrapCustom(custom);
```
Comment (Crisfole) [inline rollbar/src/main/java/com/rollbar/Rollbar.java]: This could probably be more simple:

```java
this.custom = custom == null ? null : new LinkedHashMap<String, Object>(custom);
```

File: rollbar/src/main/java/com/rollbar/Rollbar.java [hunk 4]
```diff
@@ -984,4 +986,12 @@ protected void setTransform(PayloadTransform transform) {
     public Rollbar transform(PayloadTransform transform) {
         return new Rollbar(accessToken, environment, sender, codeVersion, platform, language, framework, context, request, person, server, custom, notifier, responseHandler, filter, transform);
     }
+
+    private LinkedHashMap<String, Object> wrapCustom(Map<String, Object> custom) {
```
Comment (Crisfole) [inline rollbar/src/main/java/com/rollbar/Rollbar.java]: This is actually just a copy constructor, not really 'wrapping'.

File: rollbar/src/test/java/com/rollbar/RollbarTest.java [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+
+public class RollbarTest {
+
+    @Test
+    public void itDoesNotThrowAnExceptionWhenConfiguringWithJustAccessTokenAndEnvironment() {
```
Comment (Crisfole) [inline rollbar/src/test/java/com/rollbar/RollbarTest.java]: 👍 Thanks!

**Comment by Crisfole** [comment]
In this case there's literally nothing more descriptive since it contains all the "custom" data you want to send. Literally whatever you want.

Grep for "custom" in here: https://rollbar.com/docs/api/items_post/

**Comment by berlin-ab** [comment]
Ok, I've pushed a new commit that allows custom to be nil and just calls the copy constructor.

**Comment by Crisfole** [comment]
You rock. I'm going to pull this, run the tests, and merge it in, thanks.

**Comment by Crisfole** [comment]
And then I'm going to hook up travis...

**Comment by berlin-ab** [comment]
Great, thanks!
